include/deal.II/base/mpi.h

```diff
@@ -708,6 +709,416 @@ namespace Utilities
            const T &          object_to_send,
            const unsigned int root_process = 0);
 
+    /**
+     * An interface to be able to use the ConsensusAlgorithm classes. Main
+     * functionality of the implementations is to return a list of process ranks
+     * this process want data from and to deal with the optional payload of the
+     * messages send/received by the ConsensusAlgorithm classes.
+     *
+     * There are two kind of messages:
+     * - send/request message: A message consisting of a data request
+     *   which should be answered by an other process. This message is
+     *   considered as a request message by the receiving rank.
+     * - recv message: The answer to a send/request message.
+     *
+     * @note Since the payloads of the messages are optional, the user has
+     *       to deal with buffers himself. The ConsensusAlgorithm classes 1)
+     *       provides you only with references to empty vectors (of size 0)
+     *       to where the data to be sent can be inserted to or read from and
+     *       2) communicates these vectors blindly.
+     *
+     * @author Peter Munch, 2019
+     */
+    template <typename T1, typename T2>
+    class ConsensusAlgorithmProcess
+    {
+    public:
+      /**
+       * @return A vector of ranks to which this process wants to send a request
+       *         to.
+       *
+       * @note This is the only method which has to be implemented since the
+       *       payloads of the messages are optional.
+       */
+      virtual std::vector<unsigned int>
+      compute_targets() = 0;
+
+      /**
+       * Add to the request to the process with the specified rank a payload.
+       *
+       * @param[in] other_rank rank of the process
+       * @param[in] send_buffer data to be sent part of the request (optional)
+       *
+       * @note The buffer is empty. Before using it, you have to set its size.
+       */
+      virtual void
+      pack_recv_buffer(const int other_rank, std::vector<T1> &send_buffer);
```

Comment:
Since I assume that `send_buffer` will be populated in this function, it cannot be `const`, but then it should be marked `[in,out]` in the doxygen comment?